File: cpp/05_loops.cpp
// ============================================================
// Topic:  Loops
// Summary: C-style for, range-for, while, break/continue, transform.
// Run:     g++ -std=c++17 cpp/05_loops.cpp -o cpp/05_loops && cpp/05_loops
// ============================================================

#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

int main() {
    // ----- 1. C-style for (counter loop — like TS for (let i=0; ...)) -----
    for (int i = 0; i < 3; ++i) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
    // Output: 0 1 2

    for (int i = 2; i < 11; i += 2) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
    // Output: 2 4 6 8 10

    // ----- 2. Range-for over a collection (like Python for x in coll / TS for..of) -----
    std::vector<std::string> fruits = {"apple", "banana", "cherry"};
    for (const auto& fruit : fruits) {
        std::cout << fruit << " ";
    }
    std::cout << std::endl;
    // Output: apple banana cherry

    // ----- 3. Index + value (like Python enumerate / TS arr.entries()) -----
    for (size_t index = 0; index < fruits.size(); ++index) {
        std::cout << index << ":" << fruits[index] << " ";
    }
    std::cout << std::endl;
    // Output: 0:apple 1:banana 2:cherry

    // ----- 4. while loop -----
    int n = 3;
    while (n > 0) {
        std::cout << n << " ";
        --n;
    }
    std::cout << std::endl;
    // Output: 3 2 1

    // ----- 5. break and continue -----
    for (int i = 0; i < 10; ++i) {
        if (i == 3) continue;  // skip 3
        if (i == 6) break;     // stop at 6
        std::cout << i << " ";
    }
    std::cout << std::endl;
    // Output: 0 1 2 4 5

    // ----- 6. Transform with lambda (like Python list comp / TS arr.map) -----
    std::vector<int> nums = {0, 1, 2, 3, 4};
    std::vector<int> squares(nums.size());
    std::transform(nums.begin(), nums.end(), squares.begin(),
                   [](int x) { return x * x; });
    for (int sq : squares) {
        std::cout << sq << " ";
    }
    std::cout << std::endl;
    // Output: 0 1 4 9 16

    std::vector<int> evens;
    for (int x : nums) {
        if (x % 2 == 0) evens.push_back(x);
    }
    for (int e : evens) {
        std::cout << e << " ";
    }
    std::cout << std::endl;
    // Output: 0 2 4

    // Cross-language note:
    //   C++:     for (int i=0; i<n; ++i)   for (auto& x : vec)   while (cond)
    //   Python:  for i in range(n)         for x in coll         while cond
    //   TS:      for (let i=0; i<n; i++)    for (const x of arr)  while (cond)
    //   Rust:    for i in 0..n              for x in &vec         while cond
    //   C#:      for (int i=0; i<n; i++)    foreach (var x in list) while (cond)
    //   C++ has no for-else (Python-only). Use std::transform + lambda instead of comprehensions.
    return 0;
}

File: rust/05_loops.rs
// ============================================================
// Topic:  Loops
// Summary: loop, while, for, ranges, enumerate, break/continue.
// Run:     rustc rust/05_loops.rs -o rust/05_loops && rust/05_loops
// ============================================================

fn main() {
    // ----- 1. for over a range -----
    // 0..3 means 0, 1, 2 (exclusive end)   |  Python: range(3)   TS: for (let i=0; i<3; i++)
    print!("range 0..3: ");
    for i in 0..3 {
        print!("{} ", i); // Output: 0 1 2
    }
    println!();

    // range with step: (2..11).step_by(2)
    print!("range 2..11 step 2: ");
    for i in (2..11).step_by(2) {
        print!("{} ", i); // Output: 2 4 6 8 10
    }
    println!();

    // ----- 2. for over a collection (for..in) -----
    let fruits = ["apple", "banana", "cherry"];
    print!("fruits: ");
    for fruit in fruits.iter() {
        print!("{} ", fruit); // Output: apple banana cherry
    }
    println!();

    // ----- 3. enumerate -> index + value -----
    print!("enumerate: ");
    for (index, fruit) in fruits.iter().enumerate() {
        print!("{}:{} ", index, fruit); // Output: 0:apple 1:banana 2:cherry
    }
    println!();

    // ----- 4. while loop -----
    let mut n = 3;
    print!("while: ");
    while n > 0 {
        print!("{} ", n); // Output: 3 2 1
        n -= 1;
    }
    println!();

    // ----- 5. loop (infinite until break) with break value -----
    let mut counter = 0;
    let result = loop {
        counter += 1;
        if counter == 10 {
            break counter * 2; // break can return a value from loop
        }
    };
    println!("loop break value: {}", result); // Output: loop break value: 20

    // ----- 6. break and continue -----
    print!("break/continue: ");
    for i in 0..10 {
        if i == 3 {
            continue; // skip 3
        }
        if i == 6 {
            break; // stop at 6
        }
        print!("{} ", i); // Output: 0 1 2 4 5
    }
    println!();

    // ----- 7. Iterator methods (like Python comprehensions / TS map/filter) -----
    let squares: Vec<i32> = (0..5).map(|x| x * x).collect();
    println!("{:?}", squares); // Output: [0, 1, 4, 9, 16]

    let evens: Vec<i32> = (0..10).filter(|x| x % 2 == 0).collect();
    println!("{:?}", evens); // Output: [0, 2, 4, 6, 8]

    // Cross-language note:
    //   Rust `for x in coll` == Python `for x in coll` == TS `for (const x of coll)`.
    //   `0..n` is exclusive end; use `0..=n` for inclusive.
    //   `loop { }` has no direct Python equivalent; `while` and `for` are familiar.
    //   Only `loop` (not while/for) can `break` with a value.
}